docs/proposals/workflow.md

```diff
@@ -0,0 +1,255 @@
+`WorkflowStep` resource.
+The edges of the graph are not represented explicitly - rather they are stored as a list of
+predecessors in each `WorkflowStep` (i.e. each node).
+
+
+### Workflow
+
+A new resource will be introduced in the API. A `Workflow` is a graph.
+In the simplest case it's a graph of `Job`s but it can also
+be a graph of other entities (for example cross-cluster objects or other `Workflow`s).
+
+
+```go
+// Workflow is a directed acyclic graph
+type Workflow struct {
+    unversioned.TypeMeta `json:",inline"`
+
+    // Standard object's metadata.
+	// More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#metadata.
+	api.ObjectMeta `json:"metadata,omitempty"`
+
+    // Spec defines the expected behavior of a Workflow. More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#spec-and-status.
+    Spec WorkflowSpec `json:"spec,omitempty"`
+
+    // Status represents the current status of the Workflow. More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#spec-and-status.
+    Status WorkflowStatus `json:"status,omitempty"`
+}
+```
+
+
+#### `WorkflowSpec`
+
+```go
+// WorkflowSpec contains Workflow specification
+type WorkflowSpec struct {
+
+    // Optional duration in seconds the workflow needs to terminate gracefully. May be decreased in delete request.
+	// Value must be non-negative integer. The value zero indicates delete immediately.
+	// If this value is nil, the default grace period will be used instead.
+	// Set this value longer than the expected cleanup time for your workflow.
+    // If downstream resources (job, pod, etc.) define their TerminationGracePeriodSeconds
+    // the biggest is taken.
+	TerminationGracePeriodSeconds *int64 `json:"terminationGracePeriodSeconds,omitempty"`
+
+	// Steps contains the vertices of the workflow graph.
+	Steps []WorkflowStep `json:"steps,omitempty"`
+}
+```
+
+* `spec.steps`: is an array of `WorkflowStep`s.
+* `spec.terminationGracePeriodSeconds`: is the terminationGracePeriodSeconds.
+
+### `WorkflowStep`<sup>1</sup>
+
+The `WorkflowStep` resource acts as a [union](https://en.wikipedia.org/wiki/Tagged_union) of `JobSpec` and `ObjectReference`.
+
+```go
+// WorkflowStep contains necessary information to identifiy the node of the workflow graph
+type WorkflowStep struct {
+    // Id is the identifier of the current step
+    Id string  `json:"id,omitempty"`
+
+    // Spec contains the job specificaton that should be run in this Workflow.
+	// Only one between External and Spec can be set.
+	Spec JobSpec `json:"jobSpec,omitempty"`
+
+    // Dependecies represent dependecies of the current workflow step
+    Dependencies ObjectDependencies `json:"dependencies,omitempty"`
+
+	// External contains a reference to another schedulable resource.
+	// Only one between ExternalRef and Spec can be set.
+	ExternalRef api.ObjectReference `json:"externalRef,omitempty"`
```

Comment:
In the current proposal the user can do two things to specify the _task_ in the step:
1) via `JobSpec` (`JobTemplate` as you're proposing) [here](https://github.com/kubernetes/kubernetes/pull/18827/files#diff-bda92e294a9d99318221126e00f42f58R166): in this way you're going to create a `Job` when dependencies are satisfied.
2) via [`ExternalRef`](https://github.com/kubernetes/kubernetes/pull/18827/files#diff-bda92e294a9d99318221126e00f42f58R173): in this way the workflow step may depend on something like another workflow or something else (to address [this](https://github.com/kubernetes/kubernetes/pull/17787#issuecomment-161365446) comment).

So the _basic_ workflow use case is to have a graph of Jobs built via `JobTemplate`s if someone wants to compose workflows or use them in a cross-cluster environment it must go through `ExternalRef`.

Obviously in `ExternalRef` one could reference an already running `Job` or `ScheduledJob`.
@erictune does this answer to the question, or am I missing something?